### ISSUE TITLE
DI doesn't trigger

### ISSUE THREAD

**Issue by code-beans** [ssannandeji/Zenject-2019#60]
For some reason zenject does neither call the constructor nor the `init()`method of this Dummy class

``` csharp
public class DummyClass {

        [Inject]
        RoomController controller;

        public DummyClass() {
            Debug.Log("dummy constructor called");
        }

        [PostInject]
        void init() {
            Debug.Log("Dummyclass init called");
        }
    }
```

This is the binding

``` csharp
public class GameInstaller : MonoInstaller {

        public override void InstallBindings() {
                          Container.Bind<DummyClass>().ToSingle();  
        }
}

```

Am I doing something wrong here?

**Comment by svermeulen** [comment]
tldr: Making your class an IInitializable might fix it

My guess is that your object is not being constructed because it is not part of the initial object graph of your application.  Objects in Zenject are lazily instantiated only when they are used.  After the application is fully "installed" (ie. all installers have been called) Zenject simply instantiates one object called the Dependency Root.  Everything you want to be instantiated must be used in the object graph constructed for this root dependency.  By default it has dependencies on all IInitializable's, all ITickable's, and all IDisposables, which means that any class that you bind one of these interfaces to will always be constructed at startup.  There is also a way to explicitly add dependencies to the root as well if you want something created that doesn't derive from one of these interfaces.

In this case I think the easiest way to fix it would be to derive from IInitializable and then add this binding to your installer:

```
Container.Bind<IInitializable>().ToSingle<DummyClass>();
```

IInitializable is meant to be used for startup initialize().  [PostInject] is meant to be used for initializing dynamically created objects.

**Comment by code-beans** [comment]
I see. You might want to add this to the documentation, because it's really strange and seemed to cause non-deterministic behaviour. 

You suggestion works now.

Could you clarify on this a little bit more ? Just curious

> There is also a way to explicitly add dependencies to the root as well if you want something created that doesn't derive from one of these interfaces.

**Comment by svermeulen** [comment]
This is explained in the documentation under the Gotchas section (see "Lazily instantiated objects and the object graph").

I'm not sure what you mean by non-deterministic behaviour.  Having the object graph created based on a single dependency root seems to make sense to me also so I'm not sure what is strange about it?  

> Could you clarify on this a little bit more ? Just curious
> 
> > There is also a way to explicitly add dependencies to the root as well if you want something created that doesn't derive from one of these interfaces.

If you want to explicitly add a dependency to the initial object graph you can do something this:

```
Container.Bind<object>().ToSingle().WhenInjectedInto<IFacade>();
```

IFacade is the interface for the dependency root.  I've almost never needed to do that though, but it's there if you need it.
